### PR TITLE
Switch to new configuration framework, but keeping backward compatibility with the old one

### PR DESCRIPTION
**Pull request by marcusholl**
This change is supposed to be backward compatible.

### CODE REVIEW

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (o-liver) [inline pipelines/ui5-sap-cp/Jenkinsfile]: Shouldn't the step configuration happen inside a given step, in this case inside neoDeploy? It's because of the validation stage later, correct? How about moving the validation of each tool into the relevant step as well, removing the validation stage from the pipeline?
The drawback would be that we don't detect a missing tool early that way... Personally though I would appreciate a simpler pipeline, moving this complexity to the steps.

**Review by o-liver — COMMENTED** [review]
Can we discuss this a little.

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (marcusholl) [reply pipelines/ui5-sap-cp/Jenkinsfile]: It is a benefit to have an early validation. Otherwise long running builds are launched which fails later at the end of build, but it is clear from the very beginning that the build cannot succeed since some prerequisites are missed. In case of neoDeploy eg we fails after the build step which might have been time consuming.

Checking prerequisites (again) inside the call method of a step might be beneficial also, since the circumstances might be changed from the beginning of the pipeline (hopefully not regarding some installed tools, but there might be other things we have to check).

All in all this is a difficult thing. 
* What happens if a step is switched of, maybe based on a branch pattern (eg for pull request branches some steps are skipped)? In this case the validation must also be skipped.
* What happens if parts of the builds are executed inside a docker container? In this case it does not make sense to check the environment on the machine running jenkins.

Maybe it would be beneficial to have a validate method as part of each step which can be called early in the pipeline and also at the beginning of the call method. With that validation and step execution in the narrower sense are closer together in one place in the code. Rather than having validation code in the pipeline and step execution code in the lib. Other option would be to have a validation twin for each step (eg neoDeploy and neoDeployValidate).

More comments and ideas from all participating developers would be appreciated.

**Comment by marcusholl** [comment]
@1084565 @alejandraferreirovidal : We know about issue with validation e.g. what happens if
* parts of the pipeline runs inside docker images
* parts of the pipeline runs on nodes instead on Jenkins master
* parts of the pipeline are somehow switched off.

Now we move our code so that it is able to use also the new configuration framework (for the sake of backward compatibility we take the old one still into account). But with this we have also to update the early parts of our pipeline so that the new config framework is taken into account also here.

@1084565 The question is now - in the light of the issues mentioned above - how do we deal with the validation at all? Is it important to keep it - despite the the issues we know. Or should we remove it? Some guidance - maybe based on customer feedback - is appreciated.

**Comment by alejandraferreirovidal** [comment]
With the [new toolValidate code](https://github.com/SAP/jenkins-library/pull/75), and changing the priority of getMtaJar(), getNeoExecutable(), and getCmCliExecutable(), [see Utils](https://github.com/SAP/jenkins-library/pull/75/files#diff-1ab5d92264f5f099c7b3b5be320666c8) so MTA_JAR_LOCATION, NEO_HOME, and CM_CLI_HOME environment variables have priority over the configuration, and calling toolValidate inside the corresponding step e.g. toolValidate tool:'mta' inside the mtaBuild step, the problem should be solved.

But anyways I would prefer to finish what we are doing first, as it was planned. Of course, I still did not test this approach, and I still did not think about it much. But of course, you can not validate something that you don't know in which machine will be executed ..

**Review by alejandraferreirovidal — CHANGES_REQUESTED** [review]
I would not mix both configurations. I would keep one file with the old configuration marked as deprecated and another with the new configuration.
An so far, since it is only for testing proposes I would do it in a branch.

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (alejandraferreirovidal) [inline pipelines/ui5-sap-cp/Jenkinsfile]: This is not needed with the new config.

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (alejandraferreirovidal) [inline pipelines/ui5-sap-cp/Jenkinsfile]: And all this properties are also not needed.

Comment (alejandraferreirovidal) [reply pipelines/ui5-sap-cp/Jenkinsfile]: And all the code related to this members can be removed.

**Review by alejandraferreirovidal — CHANGES_REQUESTED** [review]
I would wait till all the PRs related to the new configuration are merged or closed to change this file.

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (marcusholl) [reply pipelines/ui5-sap-cp/Jenkinsfile]: At the moment our approach is to validated these properties. We know in the meantime that our validation does not work eg. when executing steps in dockers or on Jenkins slave nodes. But there is no decision to remove the current approach of a validation. As long as there is no such decision we validate. By the way: I personally vote for removing the validation.

**Comment by marcusholl** [comment]
Waiting until the migration to the new config framework is finished and doing a big-bang is clearly also possible. And the it is not a big thing to rework this pull request accordingly. But this means: at some point in time with the piper lib fully migrated the old pipeline (which the customer might have copied) does not work anymore. 

@1084565 What is your opinion wrt our pipeline: switch to new config framework via big-bang or with transition phase where old and new approach are working?

* pro big-bang: code is for us easier since we do not have bells and whistles for supporting both
* cons big-bang: at some point in time pipelines (customer might also work with copied pipelines) will fail in customer side.

**Comment by 1084565** [comment]
Let's stay downwards compatible for now and plan for a major version (with breaking changes) later this year.

**Comment by marcusholl** [comment]
@alejandraferreirovidal 

> With the new toolValidate code, and changing the priority

This pull request is only about introducing the new configuration framework. At some point in the past we decided to have the steps configured via the signature (... resolving the parameters first and call the steps with the parameters). It is not the aim of this pull request to change this. It is only the aim of this pull request to take the new configuration framework into account. I suggest to come to an agreement regarding regarding resolving the configuration inside the step, and to change this in another pull request (Topic of this PR is: "switch no new config framework" and not "Resolve parameters inside the steps instead of resolving it at start of pipeline"). Resolving the parameters inside the steps looks fully reasonable to me, also since we have the validation now not inside the pipeline, but inside the steps.

>I would not mix both configurations.

There is no decission at the moment about changing to a new major version (even not when this will take place). We need to be backward compatible. Here this means: we have to support both configurations.

Are there from your point of view more reasons for requesting changes?

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (alejandraferreirovidal) [inline pipelines/ui5-sap-cp/Jenkinsfile:57]: I would not keep the old configuration properties.

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (alejandraferreirovidal) [inline pipelines/ui5-sap-cp/Jenkinsfile]: I would not keep the old configuration file.

**Review by alejandraferreirovidal — CHANGES_REQUESTED** [review]

**Review by alejandraferreirovidal — CHANGES_REQUESTED** [review]
I would not keep the old configuration file. And the stage "Validate configuration" must also be adapted.

If a user is using the old pipeline, with the old configuration, it would continue working, but if a user is using this new pipeline, the user should be using the new configuration framework.

**Comment by marcusholl** [comment]
> I would not keep the old configuration file.

I disagree. Reason: backward compatibility. See also statement from @1084565. Customer changes to new pipeline, this means all projects of a customer using the pipeline needs to adjust via big-bang. With having a mixed mode we are able to handle project already on the new config and project still on the old config at the same time. IMO this is a benefit.

**Comment by alejandraferreirovidal** [comment]
Ok, as you want, but we should want users to change to the new config, not keeping both in different projects, it is not difficult to change the config files, like this they will not change the config files .. but ok.

Having different projects with different configuration frameworks is not the approach that the user should follow. They should switch to the new configuration framework.

Then the description of the PR makes no sense, because you are not switching to the new configuration, you are just adding/supporting the new configuration.

And, anyways the stage "Validate configuration" must be also adapted.

**Comment by marcusholl** [comment]
> but we should want users to change to the new config

agreed.

> it is not difficult to change the config files

You know, bigger companies, different development departments, and yet another department running central processes ... The issue is not on the technical level. The issue is more about the human factor and about interaction.

>Having different projects with different configuration frameworks is not the approach that the user should follow. They should switch to the new configuration framework.

agreed.

>They should switch to the new configuration framework.

agreed

>Then the description of the PR makes no sense, [...]

Thanks for the hint, changed.

> And, anyways the stage "Validate configuration" must be also adapted.

Thanks for the hint, I'm already working on it. First I thought about omitting since validation is shifted into steps by another pull request. But basically you are right: pull requests should be independent. And it is not much work to do it right here.

Basically I agree with a lot of your statement, the point is: backward compatibility which is considered to be a high value.

**Comment by marcusholl** [comment]
>Ok, _as you want_, 

... most of my behavior is not an attitude of arbitrariness "as I want". Most of by behavior and comments are based by rules. Here e.g. we need to be backward compatible. Of course life would be easier without being backward compatible for us. But is would make life more complicated for the users. &lt;fun&gt;The thing behind this is the _law of Conservation of Problems_ similar to the _law of Conservation of Energy_. You can shift the problems, but it is only some kind of local optimization ;-)&lt;/fun&gt;

In case you do not agree generally here, we should try to ask more colleagues.

**Comment by marcusholl** [comment]
>And, anyways the stage "Validate configuration" must be also adapted.

done, please review and suggest improvements / other approaches

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (alejandraferreirovidal) [inline pipelines/ui5-sap-cp/Jenkinsfile]: here would not be, isNewConfigFramework(CONFIG_FILE)

**Review by alejandraferreirovidal — CHANGES_REQUESTED** [review]

File: pipelines/ui5-sap-cp/Jenkinsfile
Comment (marcusholl) [reply pipelines/ui5-sap-cp/Jenkinsfile]: fixed, stupid me!

**Review by alejandraferreirovidal — APPROVED** [review]
Looks good. I did not test it.
We just have to publish the new documentation before merging this PR.

**Comment by alejandraferreirovidal** [comment]
@marcusholl the documentation is outdated. I said that I approve this PR, but the documentation should be publish before merging.

**Pull request merged**
